### ISSUE TITLE
todo: be explicit about push/pull rotation convention

### ISSUE THREAD

**Issue by alisterburt** [sjrothfuss/torch-rotate-image#8]
Something we didn't talk about but that is worth being very explicit about is what we mean by 'rotate this image n degrees in this direction'

Right handed rotations in a right handed coordinate system are expected to behave a certain way. We are currently rotating coordinates then sampling at the rotated positions, which has the effect of applying the inverse rotation. c.f. https://en.wikipedia.org/wiki/Active_and_passive_transformation

I think we should explicitly invert this so that the rotation has the same effect on the image we view as applying the rotation matrix on a set of coordinates - what do you think?

**Comment by sjrothfuss** [comment]
Yes, absolutely. Good catch. I'll be thinking about how to implement this. The first thought I have is to just "invert" the angles (360-n) before rotating coords.

**Comment by alisterburt** [comment]
Yep! that works for 2D - for 3D you'll want to invert the rotation matrices. torch.linalg.inv() will work and is cheap, I like to use it because it's explicit. Technically more efficient to transpose the rotation matrix rather than inverting it because the transpose of a rotation matrix is its inverse

**Comment by alisterburt** [comment]
worth adding a note if you implement this to explain what you're doing, it's one of these things that isn't obvious when reading code

**Comment by sjrothfuss** [comment]
Have some free time of late and resurrecting this project. @alisterburt I understand the difference between active and passive transformations and want to be sure I'm correct on the convention (since you're more aware of the field than I am). For a positive rotation angle we want a counter clockwise rotation, yes?

![Image](https://github.com/user-attachments/assets/c9c5294e-006a-4998-ad8f-49a809dfbb23)

**Comment by alisterburt** [comment]
heh, but your rendered coordinate system has y increasing as you move down...

**Comment by sjrothfuss** [comment]
If you mean the y axes in the figures I shared, I think that's an artifact of how matplotlib is plotting the arrays as images – am I misunderstanding you?

**Comment by alisterburt** [comment]
I see a CCW rotation in the image you sent but the coordinate system is flipped so that's actually a CW rotation (I think)

`origin="lower"` is the arg to imshow to sort the y-axis I think, Y axis pointing up is the same convention as 3dmod

**Comment by alisterburt** [comment]
but yes, right handed rotation around z is postive CCW when viewing XY and x is pointing right, y is pointing up

**Comment by sjrothfuss** [comment]
I understand now. Wasn't thinking about the y axis convention. If we have y increasing up, do we need to invert the rotation angle/matrix?

![Image](https://github.com/user-attachments/assets/71ff8a72-94f3-470e-b6d0-155efdc5dcc8)

**Comment by alisterburt** [comment]
got it - we have some things to fix if we want everything to be explicit 🙂 

the rotation matrices are generated here and used to rotate yx coordinates (as opposed to xy coords)

https://github.com/sjrothfuss/torch-rotate-image/blob/4593e4ef745ff31566f3f587145a4f321a317d2a/src/torch_rotate_image/rotate_image_2d.py#L59-L63

the yx convention is natural in python because image axes are ordered `(d, h, w)` and `(z, y, x)` are coordinates in each of these dimensions

looking carefully at the code used to generate the matrices shows that they are constructed as standard rotation matrices used to rotate xy coordinates

https://github.com/sjrothfuss/torch-rotate-image/blob/4593e4ef745ff31566f3f587145a4f321a317d2a/src/torch_rotate_image/rotate_image_2d.py#L87-L99

<img width="247" alt="Image" src="https://github.com/user-attachments/assets/41fcbba7-6c48-46d7-8578-7ede3e4c70e6" />

https://en.wikipedia.org/wiki/Rotation_matrix

that's the source of the discrepancy between expectations and what we observe here
